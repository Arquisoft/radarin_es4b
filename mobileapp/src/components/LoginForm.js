import React, {useState} from 'react';
import {StyleSheet, View, Text, TextInput, Linking} from 'react-native';
import {Icon} from 'react-native-elements';
import Toast from 'react-native-simple-toast';
import {authenticate} from '../api/api.js';
import getText from '../i18n.js';
import CustomLoadingButton from './CustomLoadingButton';

const LoginForm = ({changeUser}) => {
  const [idp, setIdp] = useState();
  const [username, setUsername] = useState();
  const [password, setPassword] = useState();

  const authenticateWithCredentials = async credentials => {
    let user = await authenticate(credentials);
    user ? changeUser(user) : Toast.show(getText('toastLogIn'));
  };

  return (
    <View>
      <View style={styles.container}>
        <Text style={styles.title}>{getText('logIn')}</Text>
        <Text style={styles.label}>{getText('provider')}</Text>
        <TextInput
          style={styles.input}
          onChangeText={text => setIdp(text)}
          placeholder={getText('providerPlaceholder')}></TextInput>
        <Text style={styles.label}>{getText('username')}</Text>
        <TextInput
          style={styles.input}
          onChangeText={text => setUsername(text)}
          placeholder={getText('usernamePlaceholder')}></TextInput>
        <Text style={styles.label}>{getText('password')}</Text>
        <TextInput
          style={styles.input}
          secureTextEntry={true}
          onChangeText={text => setPassword(text)}
          placeholder={getText('passwordPlaceholder')}></TextInput>
        <CustomLoadingButton
          action={() => authenticateWithCredentials({idp, username, password})}
          text={getText('enviar')}
          customContainerStyles={styles.button}
        />
      </View>
      <View style={styles.infoContainer}>
        <Icon name={'info'} />
        <Text style={styles.infoText}>{getText('loginInfo')}</Text>
        <Text style={styles.infoText}>
          {getText('getAPod')}
          <Text
            style={styles.infoLink}
            onPress={() =>
              Linking.openURL('https://solidproject.org/users/get-a-pod')
            }>
            {getText('here')}
          </Text>
          .
        </Text>
        <Text style={styles.infoText}>
          {getText('trustedApps')}
          <Text
            style={styles.infoLink}
            onPress={() =>
              Linking.openURL(
                'https://github.com/solid/userguide#manage-your-trusted-applications',
              )
            }>
            {getText('here')}
          </Text>
          .
        </Text>
      </View>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    marginStart: 'auto',
    marginEnd: 'auto',
  },
  title: {
    fontSize: 22,
    fontWeight: 'bold',
    marginBottom: 10,
    alignSelf: 'flex-start',
  },
  label: {
    fontSize: 16,
    marginBottom: 10,
    alignSelf: 'flex-start',
  },
  input: {
    height: 40,
    width: 270,
    borderColor: 'gray',
    borderWidth: 1,
    borderRadius: 4,
    marginBottom: 20,
  },
  button: {
    alignSelf: 'flex-end',
  },
  infoContainer: {
    margin: 25,
    marginTop: 30,
    padding: 15,
    backgroundColor: '#e3e3e3',
    borderRadius: 4,
    alignItems: 'flex-start',
  },
  infoText: {
    marginTop: 7,
  },
  infoLink: {
    color: '#3d48e3',
    textDecorationLine: 'underline',
  },
});

export default LoginForm;
