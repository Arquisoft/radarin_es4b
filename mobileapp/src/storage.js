import AsyncStorage from '@react-native-async-storage/async-storage';
import * as SecureStorage from 'expo-secure-store';

/**
 * Almacena un objeto con la clave especificada.
 * Si el objeto es null, borra el objeto almacenado con la clave.
 * @param {String} key clave
 * @param {*} object objeto
 */
export async function storeObject(key, object) {
  try {
    if(object) await AsyncStorage.setItem(key, JSON.stringify(object));
    else await AsyncStorage.removeItem(key);
  } catch (e) {
    console.log(e);
  }
};

/**
 * Recupera un objeto almacenado con la clave especificada
 * @param {String} key clave
 * @return {Promise<*>} Objeto almacenado 
 */
export async function getObject(key) {
  try {
    const jsonValue = await AsyncStorage.getItem(key);
    return jsonValue ? JSON.parse(jsonValue) : null;
  } catch (e) {
    console.log(e);
  }
};

/**
 * Almacena un valor con la clave especificada.
 * @param {String} key clave
 * @param {String} value valor
 */
 export async function storeValue(key, value) {
    try {
      await AsyncStorage.setItem(key, JSON.stringify(value));
    } catch (e) {
      console.log(e);
    }
  };
  
  /**
   * Recupera un valor almacenado con la clave especificada
   * @param {String} key clave
   * @return {Promise<String>} valor almacenado 
   */
  export async function getValue(key) {
    try {
      return JSON.parse(await AsyncStorage.getItem(key));
    } catch (e) {
      console.log(e);
    }
  };

/**
 * Almacena un valor sensible con la clave especificada.
 * @param {String} key clave
 * @param {String} value valor
 */
 export async function storeSecretValue(key, value) {
  try {
    await SecureStorage.setItemAsync(key, JSON.stringify(value));
  } catch (e) {
    console.log(e);
  }
};

/**
 * Recupera un valor sensible almacenado con la clave especificada
 * @param {String} key clave
 * @return {Promise<String>} valor almacenado 
 */
export async function getSecretValue(key) {
  try {
    return JSON.parse(await SecureStorage.getItemAsync(key));
  } catch (e) {
    console.log(e);
  }
};
