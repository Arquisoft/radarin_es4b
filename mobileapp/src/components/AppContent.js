import React, {useEffect, useState, useRef} from 'react';
import {ScrollView, View} from 'react-native';
import LoginForm from './LoginForm';
import {getObject, storeObject} from '../storage.js';
import LoggedInView from './LoggedInView.js';
import * as CurrentUser from '../user.js';
import SplashScreen from 'react-native-splash-screen';
import {getSecretValue} from '../storage.js';
import {hashCode} from '../utils.js'

const AppContent = () => {
  const [user, setUser] = useState();
  const didMount = useRef(false);

  useEffect(() => {
    getObject('user').then(user => {
      if (user) {
        getSecretValue(`${hashCode(user.webId)}-token`).then(token => {
          if (token) {
            CurrentUser.setToken(token);
            setUser(user);
            return;
          }
        });
      }
      SplashScreen.hide();
    });
  }, []);

  useEffect(() => {
    if (didMount.current) {
      if (user) SplashScreen.hide();
      storeObject('user', user);
      CurrentUser.setWebId(user?.webId);
    } else didMount.current = true;
  }, [user]);

  return (
    <View
      style={{
        marginTop: 30,
        flex: 1,
      }}>
      {!user && (
        <ScrollView>
          <LoginForm changeUser={setUser}></LoginForm>
        </ScrollView>
      )}
      {user && <LoggedInView user={user} changeUser={setUser}></LoggedInView>}
    </View>
  );
};

export default AppContent;
