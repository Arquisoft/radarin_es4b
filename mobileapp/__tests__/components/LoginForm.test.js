/**
 * @format
 */

 import 'react-native';
 import React from 'react';
 import getText from '../../src/i18n.js';
 import {storeObject} from '../../src/storage.js';
 
 // Note: test renderer must be required after react-native.
 import renderer from 'react-test-renderer';
 import {render, fireEvent, waitFor} from '@testing-library/react-native';
 import LoginForm from '../../src/components/LoginForm';
 import AppContent from '../../src/components/AppContent';
 
 const expectUser = {
    token: 'tokenDelUsuario',
    webId: 'https://davidaf.solidcommunity.net/profile/card#me',
    name: 'David Álvarez',
    photo: 'https://davidaf.solidcommunity.net/profile/fotosolid.png',
 };

 test('log in form renders correctly', () => {
    const changeUser = user => {
      console.log(user);
    };
  
    const logInForm = renderer
      .create(<LoginForm changeUser={changeUser}></LoginForm>)
      .toJSON();
    expect(logInForm).toMatchSnapshot();
  });
  
  test('user with valid credentials can log in', async () => {
    fetch.mockResponseOnce(JSON.stringify(expectUser), {status: 200});

    await storeObject('user', null);
  
    const {getByText, getByPlaceholderText} = render(<AppContent />);
    await waitFor(() => {
      getByText(getText('enviar'));
    });
  
    fireEvent.changeText(
      getByPlaceholderText(getText('providerPlaceholder')),
      'https://solidcommunity.com',
    );
    fireEvent.changeText(
      getByPlaceholderText(getText('usernamePlaceholder')),
      'david',
    );
    fireEvent.changeText(
      getByPlaceholderText(getText('passwordPlaceholder')),
      '1234',
    );
    fireEvent.press(getByText(getText('enviar')));
  
    // Comprueba que el usuario accede a la vista de usuario logeado
    await waitFor(() => {
      getByText(expectUser.name);
    });
  });
  
  test('user with invalid credentials cant log in', async () => {
    fetch.mockResponseOnce('Error', {status: 403});
  
    await storeObject('user', null);
  
    const {getByPlaceholderText, getByText} = render(
      <AppContent />,
    );
    fireEvent.changeText(
      getByPlaceholderText(getText('providerPlaceholder')),
      'https://provider.com',
    );
    fireEvent.changeText(
      getByPlaceholderText(getText('usernamePlaceholder')),
      'Pepito',
    );
    fireEvent.changeText(
      getByPlaceholderText(getText('passwordPlaceholder')),
      '7777',
    );
  
    fireEvent.press(getByText(getText('enviar')));
  
    // Comprueba que no ha pasado ha la vista de usuario logeado
    await waitFor(() => {
      getByText(getText('enviar'));
    });
  });