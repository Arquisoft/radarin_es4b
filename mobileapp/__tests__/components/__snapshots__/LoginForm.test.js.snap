// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`log in form renders correctly 1`] = `
<View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "marginEnd": "auto",
        "marginStart": "auto",
      }
    }
  >
    <Text
      style={
        Object {
          "alignSelf": "flex-start",
          "fontSize": 22,
          "fontWeight": "bold",
          "marginBottom": 10,
        }
      }
    >
      Log in
    </Text>
    <Text
      style={
        Object {
          "alignSelf": "flex-start",
          "fontSize": 16,
          "marginBottom": 10,
        }
      }
    >
      Solid identity provider:
    </Text>
    <TextInput
      allowFontScaling={true}
      onChangeText={[Function]}
      placeholder="e. g. https://solidcommunity.net"
      rejectResponderTermination={true}
      style={
        Object {
          "borderColor": "gray",
          "borderRadius": 4,
          "borderWidth": 1,
          "height": 40,
          "marginBottom": 20,
          "width": 270,
        }
      }
      underlineColorAndroid="transparent"
    />
    <Text
      style={
        Object {
          "alignSelf": "flex-start",
          "fontSize": 16,
          "marginBottom": 10,
        }
      }
    >
      Username:
    </Text>
    <TextInput
      allowFontScaling={true}
      onChangeText={[Function]}
      placeholder="Username"
      rejectResponderTermination={true}
      style={
        Object {
          "borderColor": "gray",
          "borderRadius": 4,
          "borderWidth": 1,
          "height": 40,
          "marginBottom": 20,
          "width": 270,
        }
      }
      underlineColorAndroid="transparent"
    />
    <Text
      style={
        Object {
          "alignSelf": "flex-start",
          "fontSize": 16,
          "marginBottom": 10,
        }
      }
    >
      Password:
    </Text>
    <TextInput
      allowFontScaling={true}
      onChangeText={[Function]}
      placeholder="Password"
      rejectResponderTermination={true}
      secureTextEntry={true}
      style={
        Object {
          "borderColor": "gray",
          "borderRadius": 4,
          "borderWidth": 1,
          "height": 40,
          "marginBottom": 20,
          "width": 270,
        }
      }
      underlineColorAndroid="transparent"
    />
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignSelf": "flex-end",
        }
      }
    >
      <Text
        style={
          Object {
            "alignSelf": "flex-end",
            "backgroundColor": "#007bff",
            "borderRadius": 4,
            "color": "white",
            "fontSize": 16,
            "height": 40,
            "textAlign": "center",
            "textAlignVertical": "center",
            "width": 100,
          }
        }
      >
        Send
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "flex-start",
        "backgroundColor": "#e3e3e3",
        "borderRadius": 4,
        "margin": 25,
        "marginTop": 30,
        "padding": 15,
      }
    }
  >
    <View
      style={
        Object {
          "overflow": "hidden",
        }
      }
    >
      <View>
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "transparent",
              "justifyContent": "center",
            }
          }
        >
          <Text
            allowFontScaling={false}
            style={
              Array [
                Object {
                  "color": "#242424",
                  "fontSize": 24,
                },
                Object {
                  "backgroundColor": "transparent",
                },
                Object {
                  "fontFamily": "Material Icons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
            testID="iconIcon"
          >
            î¢Ž
          </Text>
        </View>
      </View>
    </View>
    <Text
      style={
        Object {
          "marginTop": 7,
        }
      }
    >
      To use Radarin you need to have a Solid POD and have added https://solid-node-client as a trusted app
    </Text>
    <Text
      style={
        Object {
          "marginTop": 7,
        }
      }
    >
      If you don't have a POD, you can get one 
      <Text
        onPress={[Function]}
        style={
          Object {
            "color": "#3d48e3",
            "textDecorationLine": "underline",
          }
        }
      >
        here
      </Text>
      .
    </Text>
    <Text
      style={
        Object {
          "marginTop": 7,
        }
      }
    >
      If you don't know how to add a trusted app, you can find more information 
      <Text
        onPress={[Function]}
        style={
          Object {
            "color": "#3d48e3",
            "textDecorationLine": "underline",
          }
        }
      >
        here
      </Text>
      .
    </Text>
  </View>
</View>
`;
